[WTF] Drop DisallowCType.h

sthibaul opened the pull request:
#### f817cd0c3d7044500db7bdeb43be96a539d8595c
<pre>
[WTF] Drop DisallowCType.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=285806">https://bugs.webkit.org/show_bug.cgi?id=285806</a>

Reviewed by Michael Catanzaro.

DisallowCType.h was meant to make sure that WebKit does not use the
locale-dependent ctype functions for locale-independent purpose.

The git log of this file however shows that this breaks in a lot of cases
(musl, darwin, glibc regex.h, etc.), which were worked around until a point
where it is actually hard to find a system where this check would actually
happen, so better just drop it.

This fixes GNU/Hurd build whose glibc also had regex.h include ctype
like on GNU/Linux.

* Source/JavaScriptCore/config.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/config.h:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/DisallowCType.h: Removed.
* Source/WebCore/PAL/config.h:
* Source/WebCore/config.h:
* Source/WebDriver/config.h:
* Source/WebGPU/WGSL/config.h:
* Source/WebGPU/WebGPU/config.h:
* Source/WebKit/WPEPlatform/config.h:
* Source/WebKit/config.h:
* Source/WebKitLegacy/mac/WebKitPrefix.h:
* Tools/TestWebKitAPI/config.h:

Canonical link: <a href="https://commits.webkit.org/288854@main">https://commits.webkit.org/288854@main</a>
</pre>

mcatanzaro commented on Source/WTF/wtf/DisallowCType.h:
```diff
@@ -41,7 +41,7 @@
 // or <glib/gi18n-lib.h>, which in turn include <xlocale/_ctype.h> which uses
 // isacii(). Additionally, regex.h uses tolower on Linux.
-#if !OS(LINUX) && !defined(_LIBCPP_VERSION) && defined(__GLIBC__)
+#if !OS(LINUX) && !OS(HURD) && !defined(_LIBCPP_VERSION) && defined(__GLIBC__)
```
Hmmm, I'm thinking we need a strategy rethink here. A glibc-specific header that's not used on Linux doesn't make sense. You might as well delete this header entirely. This comes from https://commits.webkit.org/170756@main which we might need to reconsider.

sthibaul replied:
The addition of `defined(__GLIBC__)` in  https://commits.webkit.org/170756@main  comes from a time when linux was not excluded, so it did somehow make sense at the time, though it was an odd way to say "don't try this on musl". AIUI we'd want to include this header to make some checks on some OS which permit to do this check, but it looks like it's a lost case when apparently so many things go against this approach. I'll thus submit the removal.

mcatanzaro replied:
CC: Darin since he added this header in https://commits.webkit.org/20634@main

darinadler replied:
This was never intended to be glib-specific. If we can make it work, this is still useful for the same reason I originally added it. On platforms where it _does_ compile, it catches us if we accidentally use `ctype` functions, which are never a good idea in WebKit since there are platforms where `ctype` is affected by the `LOCALE` environment. I will reluctantly agree to delete it if we can't get it to work anywhere. I'm a bit irritated that it was turned off for Apple platforms and I don't remember when we did that.

mcatanzaro replied:
It's currently specific to glibc (GNU C library), not glib (GNOME library).

sthibaul replied:
The `__GLIBC__` condition was added only to exclude musl, and not really meant to exclude apple. And later on Linux got excluded by @zdobersek , I don't see why actually, the error I'm getting if I drop that condition is

```
/usr/include/c++/14/bits/chrono_io.h: In member function ‘std::basic_istream<_CharT2, _Traits2>& std::chrono::__detail::_Parser<_Duration>::operator()(std::basic_istream<_CharT2, _Traits2>&, const _CharT*, std::__cxx11::basic_string<_CharT2, _Traits2, _Alloc>*, std::chrono::minutes*)’:
/build/webkit2gtk-2.46.5/Source/WTF/wtf/DisallowCType.h:75:17: error: ‘isspace_WTF_Please_use_ASCIICType_instead_of_ctype_see_comment_in_ASCIICType_h’ is not a member of ‘std’
   75 | #define isspace isspace_WTF_Please_use_ASCIICType_instead_of_ctype_see_comment_in_ASCIICType_h
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Possibly, instead of letting various people try to add their own if !(FOO), we could just want to turn that into 

#if OS(DARWIN) && !PLATFORM(GTK) && !PLATFORM(WPE) && !defined(_LIBCPP_VERSION)

so it gets tested there for good without risking getting dropped?

sthibaul replied:
I have changed the PR so, let's see the builds

mcatanzaro commented on Source/WTF/wtf/DisallowCType.h:
```diff
@@ -41,7 +41,7 @@
 // or <glib/gi18n-lib.h>, which in turn include <xlocale/_ctype.h> which uses
 // isacii(). Additionally, regex.h uses tolower on Linux.
-#if !OS(LINUX) && !defined(_LIBCPP_VERSION) && defined(__GLIBC__)
+#if OS(DARWIN) && !PLATFORM(GTK) && !PLATFORM(WPE) && !defined(_LIBCPP_VERSION)
```
This is unfortunately also pointless because Apple uses libc++. We'll never build with `OS(DARWIN)` but not `defined(_LIBCPP_VERSION)`.

I think it's totally fine to prevent WebKit's own code from using these functions, but the current strategy does not work because it affects system headers as well. My suggestion: accept the previous version of this pull request that removes this header altogether, and instead add a style checker rule that prohibits use of these functions.

sthibaul replied:
arf... Ok, reverted to the previous version.

mcatanzaro replied:
@darinadler are you OK with this plan?

darinadler replied:
Our goal is to get a compilation failure if we accidentally use `ctype` functions. I don’t see any comment explaining why this fails on Apple platforms: all the comments seem to be about other platforms.

If we don’t know how to do that successfully even on the Apple platforms, then I will accept that we failed, and just can’t do it and I am fine with a style checker rule instead if it works in practice. But when we say “the current strategy does not work because it affects system headers as well”, did we try, and fail, to address that? Originally I was able to get this to work, but I understand if it’s not been practical since then.

mcatanzaro replied:
Well https://github.com/WebKit/WebKit/pull/38909#discussion_r1913811265 demonstrates failure with libstdc++, there's just no way to include chrono because we've undefined stuff that it uses.

Then it's been disabled when building with libc++ since https://commits.webkit.org/112146@main which references https://bugs.webkit.org/show_bug.cgi?id=93849 and shows a similar error:
 
 ```
 /home/milian/bbndk-2.0.1/target/qnx6/usr/include/cpp/xlocinfo: In function 'int std::_Tolower(int, const std::_Ctypevec*)':
/home/milian/bbndk-2.0.1/target/qnx6/usr/include/cpp/xlocinfo:173: error: 'tolower_WTF_Please_use_ASCIICType_instead_of_ctype_see_comment_in_ASCIICType_h' is not a member of 'std'
/home/milian/bbndk-2.0.1/target/qnx6/usr/include/cpp/xlocinfo: In function 'int std::_Toupper(int, const std::_Ctypevec*)':
/home/milian/bbndk-2.0.1/target/qnx6/usr/include/cpp/xlocinfo:179: error: 'toupper_WTF_Please_use_ASCIICType_instead_of_ctype_see_comment_in_ASCIICType_h' is not a member of 'std'
```

Presumably it has been disabled for Apple builds ever since Apple switched from libstdc++ to libc++, which I assume was probably about a decade ago? You did use libstdc++ before LLVM/Clang came onto the scene, right?

darinadler replied:
OK, understood. This hasn't been working for a while in practice.

mcatanzaro reviewed (approved):
OK, I'll approve this.

I checked the style checker and unfortunately I couldn't find an existing list of disallowed function names, so we'd need to implement a new check rule to do this. Let's not block on that. If you could just file a bug report against the Tools component so that we have an open bug to track the problem, that would be great.

sthibaul committed:
[WTF] Drop DisallowCType.h
https://bugs.webkit.org/show_bug.cgi?id=285806

Reviewed by Michael Catanzaro.

DisallowCType.h was meant to make sure that WebKit does not use the
locale-dependent ctype functions for locale-independent purpose.

The git log of this file however shows that this breaks in a lot of cases
(musl, darwin, glibc regex.h, etc.), which were worked around until a point
where it is actually hard to find a system where this check would actually
happen, so better just drop it.

This fixes GNU/Hurd build whose glibc also had regex.h include ctype
like on GNU/Linux.

* Source/JavaScriptCore/config.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/config.h:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/DisallowCType.h: Removed.
* Source/WebCore/PAL/config.h:
* Source/WebCore/config.h:
* Source/WebDriver/config.h:
* Source/WebGPU/WGSL/config.h:
* Source/WebGPU/WebGPU/config.h:
* Source/WebKit/WPEPlatform/config.h:
* Source/WebKit/config.h:
* Source/WebKitLegacy/mac/WebKitPrefix.h:
* Tools/TestWebKitAPI/config.h:

Canonical link: https://commits.webkit.org/288854@main
